### PR TITLE
LX-1551, LX-1553 - dx_apply & dx_execute 

### PR DESCRIPTION
**Pull request by sdimitro**
LX-1551 dx_apply for migration
    LX-1553 dx_execute for migration

    This patch introduces the next 2 migration scripts:
    dx_apply and dx_execute.

    dx_apply: creates the dataset hierarchy expected by linux upgrade
    and unpacks the migration image within it. It also adds a new
    entry in the FreeBSD bootloader so a VM can optionally boot from
    Linux. Note that the script has a few WIP comments that are there
    for 2 WIP issues. The first one is the fact that we can't do any
    checks about running on the 5.3 release because of the feature
    incompatibility of ZoL and our illumos pools (this is tracked by
    LX-1633). The other one is a placeholder for the point of execution
    where the migration scripts will run.

    dx_execute: checks that dx_apply has left around all the expected
    files and info and configures the bootloader to boot into Linux
    next time the system boots.

    Side-note - This patch also fixes issue #170.

    Testing:

    In a DRYRUN-5.3.2.0 VM:
    ```
    $ sudo /opt/delphix/server/bin/upgrade/dx_unpack internal-qa.migration.tar.gz
    Progress increment: 17:52:10:272024270+0000, 10, Extracting upgrade image
    17:52:10:274636821:+0000: Unpacking internal-qa.migration.tar.gz ...

    18:02:32:771501160:+0000: done.
    Progress increment: 18:02:32:774283991+0000, 40,  verifying format
    Progress increment: 18:02:32:802230044+0000, 50,  verifying signature
    18:02:32:804885537:+0000: Verify signature ... Verified OK
    Progress increment: 18:02:32:940595417+0000, 70, Verifying integrity of upgrade image
    18:02:32:943585708:+0000: Verifying contents of ./dx_apply ... ./dx_apply: OK
    18:02:32:959290157:+0000: Verifying contents of ./dx_execute ... ./dx_execute: OK
    18:02:32:973580872:+0000: Verifying contents of ./dx_prepare ... ./dx_prepare: OK
    18:02:32:988391332:+0000: Verifying contents of ./dx_verify ... ./dx_verify: OK
    18:02:33:005307726:+0000: Verifying contents of ./os-root.cpio ... ./os-root.cpio: OK
    18:04:28:832789838:+0000: Verifying contents of ./os-root.hashes ... ./os-root.hashes: OK
    18:04:28:970944902:+0000: Verifying contents of ./version.info ... ./version.info: OK
    Progress increment: 18:04:28:986276858+0000, 80, Verifying integrity of upgrade image done
    Progress increment: 18:04:29:062561886+0000, 90, preparing upgrade image
    Progress increment: 18:04:29:078086553+0000, 100, unpacking successful

    $ ls /var/dlpx-update/latest/
    dx_apply  dx_execute  dx_prepare  dx_verify  hashes  hashes.sig.5.3  os-root.cpio  os-root.hashes  version.info

    $ sudo /var/dlpx-update/latest/dx_apply /var/dlpx-update/latest/
    28115061 blocks

    $ zfs list
    NAME                                                      USED  AVAIL  REFER  MOUNTPOINT
    rpool                                                    42.7G   102G  43.5K  /rpool
    rpool/ROOT                                               9.36G   102G    23K  none
    rpool/ROOT/delphix.C3zxml7                               9.36G   102G  9.36G  /
    ...

    $ ls /boot/
    ...  initrd.img-4.15.0-42-generic ...
    ... vmlinuz-4.15.0-42-generic ...

    $ tail /boot/menu.rc.local
    set dversionsmenu_caption[3]="boot archive recovery"
    set dv_ba[3]="-B dlpx_version=5.3.2.0,dlpx_ds=rpool/versions/5.3.2018.11.21/dlpx/5.3.2.0,dlpx_support=true"
    set dversionsmenu_command[3]="s\"  set currdev=zfs:rpool/versions/5.3.2018.11.21/support:\" evaluate s\" set  boot-args=\${dv_ba[3]}\" evaluate boot"
    set mainmenu_caption[8]="Boot Delphix on [L]inux"
    set mainansi_caption[8]="Boot Delphix on ^[1mL^[minux version"
    set mainmenu_keycode[8]=76
    set mainmenu_command[8]="s\" load /boot/vmlinuz-4.15.0-42-generic root=ZFS=rpool/ROOT/delphix.C3zxml7 console=tty0 console=ttyS0,115200n8 zfsforce=1\" evaluate s\" load -t rootfs /boot/initrd.img-4.15.0-42-generic\" evaluate boot"

    $ sudo /var/dlpx-update/latest/dx_execute
    $

    $ tail /boot/menu.rc.local
    set dversionsmenu_caption[3]="boot archive recovery"
    set dv_ba[3]="-B dlpx_version=5.3.2.0,dlpx_ds=rpool/versions/5.3.2018.11.21/dlpx/5.3.2.0,dlpx_support=true"
    set dversionsmenu_command[3]="s\"  set currdev=zfs:rpool/versions/5.3.2018.11.21/support:\" evaluate s\" set  boot-args=\${dv_ba[3]}\" evaluate boot"
    set mainmenu_caption[8]="Boot Delphix on [L]inux"
    set mainansi_caption[8]="Boot Delphix on ^[1mL^[minux version"
    set mainmenu_keycode[8]=76
    set mainmenu_command[8]="s\" load /boot/vmlinuz-4.15.0-42-generic root=ZFS=rpool/ROOT/delphix.C3zxml7 console=tty0 console=ttyS0,115200n8 zfsforce=1\" evaluate s\" load -t rootfs /boot/initrd.img-4.15.0-42-generic\" evaluate boot"
    set menu_timeout_command="s\" load /boot/vmlinuz-4.15.0-42-generic root=ZFS=rpool/ROOT/delphix.C3zxml7 console=tty0 console

    $ sudo reboot

    <....sshing back in after some time ...>

    $ uname -a
        Linux localhost 4.15.0-42-generic #45-Ubuntu SMP Thu Nov 15 19:32:57 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
    ```

    After rebooting the VM ~3 times from within Linux, we get into the recovery
    environment:
    ```
    Please enter a number  [1]:
                    Welcome to the Delphix Server 5.3.2.0 Installation Menu

            1  Shell
            2  Terminal type (currently sun-color)
            3  Reboot
    ```

ab-pre-push (pending):
http://selfservice.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-orchestrator-pre-push/335/flowGraphTable/

### CODE REVIEW

File: live-build/misc/migration-scripts/dx_apply [hunk 1]
```diff
@@ -16,9 +16,190 @@
...
+#
+RDS=$(mount | awk '/^\/ /{ print $3 }')
+RPOOL=${RDS%%/*}
+
+#
+# Cleanup any previous intermediate state.
+#
+rm -rf /tmp/delphix.*
+rm -f /boot/vmlinuz-* /boot/initrd.img-*
+zfs destroy -r "$RPOOL/ROOT" 2>/dev/null
```
Comment (pzakha) [inline live-build/misc/migration-scripts/dx_apply]: We should probably also double check that "$RPOOL/ROOT" doesn't exist after this operation, since we don't check for the failure of `zfs destroy`.

File: live-build/misc/migration-scripts/dx_apply [hunk 2]
```diff
@@ -16,9 +16,190 @@
...
+# - We specify zfsforce=1 because otherwise we would be failing at
+#   pool import because of ZFS's hostname check.
+#
+LOAD_VMLINUZ_OK_ARG=(
+	"/boot/$(basename "$VMLINUZ")"
+	"root=ZFS=$RPOOL/ROOT/$FSNAME"
+	'console=tty0 console=ttyS0,115200n8'
+	'zfsforce=1'
+)
+LOAD_VMLINUZ_OK_CMD="load ${LOAD_VMLINUZ_OK_ARG[*]}"
```
Comment (pzakha) [inline live-build/misc/migration-scripts/dx_apply]: Any particular reasons for using an array here instead of a regular string? It also seems like quote types `'` and `"` are mixed in the arguments for `LOAD_VMLINUZ_OK_ARG`.

File: live-build/misc/migration-scripts/dx_apply [hunk 3]
```diff
@@ -16,9 +16,190 @@
...
+
+#
+# This command loads the initial RAM disk as the initial root filessytem
```
Comment (pzakha) [inline live-build/misc/migration-scripts/dx_apply]: typo: filessytem -> filesystem

File: live-build/misc/migration-scripts/dx_apply [hunk 4]
```diff
@@ -16,9 +16,190 @@
...
+# bootloader menu is chosen. It is FICL code (a Forth-inspired
+# command language) that is basically read as:
+# 	arg func; arg func; func.
+# In our case we:
+# [1] Evaluate the command to load vmlinuz
+# [2] Evaluate the command to load initrd
+# [3] boot (with the above two loaded)
+#
+MAIN_MENU_FICL=(
+	's\"' $LOAD_VMLINUZ_OK_CMD'\"' evaluate
```
Comment (pzakha) [inline live-build/misc/migration-scripts/dx_apply]: I don't know much about `forth`, so it's more of a question:
Why do we need the 's\"' stuff and the evaluate. cannot we just pass:
```
<arg-to-load> load <arg-to-2nd-load> load boot
```
Also don't we need any semi-colons between the commands?

Looking at the output you've posted in the PR description, it seems like the escaping sequences for those 2 commands are different:
```
set mainmenu_command[8]="s\" load /boot/vmlinuz-4.15.0-42-generic root=ZFS=rpool/ROOT/delphix.C3zxml7 console=tty0 console=ttyS0,115200n8 zfsforce=1\" evaluate s\" load -t rootfs /boot/initrd.img-4.15.0-42-generic\" evaluate boot"
set menu_timeout_command="s" load /boot/vmlinuz-4.15.0-42-generic root=ZFS=rpool/ROOT/delphix.vI9uLxo console=tty0 console=ttyS0,115200n8 zfsforce=1" evaluate s" load -t rootfs /boot/initrd.img-4.15.0-42-generic" evaluate boot"
```
One of them has `"s\"`, the other one has `"s"`

File: live-build/misc/migration-scripts/dx_execute [hunk 5]
```diff
@@ -16,9 +16,71 @@
...
+# CURRENT_DDS=$(dirname "$(mount | awk '/^\/opt\/delphix /{ print $3 }')")
+# CURRENT_VERSION=$(basename $current_dds)
+#
+# WIP: check if current version is 5.3, if not bail
+#
+
+#
+# Ensure that the expected Linux dataset layout exists.
+#
+[[ $(zfs list -o name -Hr rpool/ROOT | wc -l) -eq 2 ]] ||
```
Comment (pzakha) [inline live-build/misc/migration-scripts/dx_execute]: `rpool/ROOT` -> `"$RPOOL/ROOT"`
We should stay consistent. I'm wondering if we shouldn't just use `rpool` everywhere. I can't think of situations where the root pool is not called `rpool`.

File: live-build/misc/migration-scripts/dx_execute [hunk 6]
```diff
@@ -16,9 +16,71 @@
...
+	die "could not find the expected linux dataset layout"
+LX_RPOOL_ROOT=$(zfs list -o name -H "$RPOOL/ROOT")
+[[ -n $LX_RPOOL_ROOT ]] || die "could not find Linux ROOT dataset"
+LX_RDS=$(zfs list -o name -Hr "$RPOOL/ROOT" | tail -n 1)
+
+#
+# Ensure that the expected bootloader fields are there.
+#
+[[ "$(grep -c 'mainmenu_caption\[8\]' /boot/menu.rc.local)" -eq 1 ]] ||
```
Comment (pzakha) [inline live-build/misc/migration-scripts/dx_execute]: hint: you can add -F to grep and remove character escaping

File: live-build/misc/migration-scripts/dx_apply [hunk 7]
```diff
@@ -16,9 +16,190 @@
...
+	's\"' $LOAD_INITRD_OK_CMD'\"' evaluate
```
Comment (pzakha) [inline live-build/misc/migration-scripts/dx_apply]: Not sure if I have a different version of `shellcheck` on my mac from the one run by the PR automation, but while the PR checks seem to all pass, mine complain about the escaping in those 2 lines:
```
In live-build/misc/migration-scripts/dx_apply line 171:
	's\"' $LOAD_VMLINUZ_OK_CMD'\"' evaluate
              ^-- SC2206: Quote to prevent word splitting, or split robustly with mapfile or read -a.


In live-build/misc/migration-scripts/dx_apply line 172:
	's\"' $LOAD_INITRD_OK_CMD'\"' evaluate
              ^-- SC2206: Quote to prevent word splitting, or split robustly with mapfile or read -a.
```

File: live-build/misc/migration-scripts/dx_apply [hunk 8]
```diff
@@ -16,9 +16,190 @@
...
+done
+
+ARCHIVE_DIR=$1
+[[ -n $ARCHIVE_DIR ]] || usage
+[[ -d $ARCHIVE_DIR ]] || die "$ARCHIVE_DIR is not a directory"
+[[ -f $ARCHIVE_DIR/version.info ]] ||
+	die "$ARCHIVE_DIR does not have a version.info file"
+. "$ARCHIVE_DIR/version.info"
+#
+# WIP: Assert MINIMUM VERSION field for migration.
```
[the remark this answers is not in the document]
Comment (sdimitro) [reply live-build/misc/migration-scripts/dx_apply]: Yes, so there are basically 2 checks about the minimum version.

The first one happens right after we source `version.info` from the migration archive:
```
 52 . "$ARCHIVE_DIR/version.info"
 53 #
 54 # WIP: Assert MINIMUM VERSION field for migration.
 55 #
```
This is to ensure that the archive itself provides the correct minimum version. The idea is to guard against this migration script used against some random image that is for example just upgrade and not migration.

The second one is looking for what is currently installed on the system, and making sure that the system is at that minimum version:
```
 74 #
 75 # Check that the version to which we are upgrading is compatible with the
 76 # currently installed version. (Look at version.info to see if we are
 77 # above the minimum version required. Since we are doing a migration
 78 # the minimum version should be the latest 5.3.
 79 #
 80 # CURRENT_DDS=$(dirname "$(mount | awk '/^\/opt\/delphix /{ print $3 }')")
 81 # CURRENT_VERSION=$(basename $current_dds)
 82 #
 83 # WIP: check if current version is 5.3, if not bail
 84 #
```

TL;DR - First check is to trying to check if we have the right archive. Second check ensures that the system is at the minimum version.

Does this sound reasonable? If yes, I can write more comments to clarify. If not, I can just take the comments out of there.

File: live-build/misc/migration-scripts/dx_apply [hunk 9]
```diff
@@ -16,9 +16,190 @@
...
+[[ "$(uname -s)" == "SunOS" ]] || die "script can only be used in illumos"
+
+while getopts ':h' opt; do
+	case $opt in
+	h) usage ;;
+	*) usage "Invalid option: -$OPTARG." ;;
+	esac
+done
+
+ARCHIVE_DIR=$1
```
[the remark this answers is not in the document]
Comment (sdimitro) [reply live-build/misc/migration-scripts/dx_apply]: new revision is double-checked with local MacOS shellcheck too

File: live-build/misc/migration-scripts/dx_apply (hunk 4 above)
Comment (sdimitro) [reply live-build/misc/migration-scripts/dx_apply]: As mentioned in our conversation, these are Forth strings (https://www.complang.tuwien.ac.at/forth/gforth/Docs-html/Characters-and-Strings-Tutorial.html) evaluated within forth.

As for the output line without the slash that was a presentation/editing fault on my part.

File: live-build/misc/migration-scripts/dx_apply (hunk 7 above)
Comment (sdimitro) [reply live-build/misc/migration-scripts/dx_apply]: I changed that to have a single whole string per argument with double quotes.
The only awkward thing is that if you want something like `\"` you need to escape both the slash and the quote, leading in -> `\\\"`. After looking at this for a while that seemed nicer to me than mixing and matching single and double quotes.

File: live-build/misc/migration-scripts/dx_apply [hunk 10]
```diff
@@ -16,9 +16,220 @@
...
+# Set things up for the FreeBSD bootloader to provide an option for
+# booting into the Linux environment This happens in two steps:
+# [1] We place vmlinuz and initrd from our migration image into
+#     the /boot directory of the current system, so the bootloader
+#     can load them.
+# [2] Add a new option in the bootloader's menu for Linux. This is
+#     done by appending the configuration of the new option in our
+#     current menu file.
+#
+VMLINUZ=$(find "$TMP_ROOT/boot/vmlinuz-"* | sort -r | head -n 1)
```
[the remark this answers is not in the document]
Comment (sdimitro) [reply live-build/misc/migration-scripts/dx_apply]: Comment added

**Comment by prakashsurya** [comment]
bors delegate+
